rosetta/retriever/retriever_test.go

```diff
@@ -0,0 +1,1196 @@
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				return identifier.Block{}, errors.New("invalid block")
+			},
+		}
+		r := &Retriever{
+			validate: validator,
+		}
+
+		_, _, err := r.Balances(testBlockID, testAccount, testCurrencies)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles invalid account", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) { return block, nil },
+			AccountFunc: func(address identifier.Account) error {
+				return errors.New("invalid account")
+			},
+		}
+		r := &Retriever{
+			validate: validator,
+		}
+
+		_, _, err := r.Balances(testBlockID, testAccount, testCurrencies)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles invalid currency", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc:   func(block identifier.Block) (identifier.Block, error) { return block, nil },
+			AccountFunc: func(address identifier.Account) error { return nil },
+			CurrencyFunc: func(currency identifier.Currency) (identifier.Currency, error) {
+				return currency, errors.New("invalid currency")
+			},
+		}
+		r := &Retriever{
+			validate: validator,
+		}
+
+		_, _, err := r.Balances(testBlockID, testAccount, testCurrencies)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles generator failure", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc:    func(block identifier.Block) (identifier.Block, error) { return block, nil },
+			AccountFunc:  func(address identifier.Account) error { return nil },
+			CurrencyFunc: func(currency identifier.Currency) (identifier.Currency, error) { return currency, nil },
+		}
+		generator := &mocks.Generator{
+			GetBalanceFunc: func(symbol string) ([]byte, error) {
+				return nil, mocks.DummyError
+			},
+		}
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+		}
+
+		_, _, err := r.Balances(testBlockID, testAccount, testCurrencies)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles invoker failure", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc:    func(block identifier.Block) (identifier.Block, error) { return block, nil },
+			AccountFunc:  func(address identifier.Account) error { return nil },
+			CurrencyFunc: func(currency identifier.Currency) (identifier.Currency, error) { return currency, nil },
+		}
+		generator := &mocks.Generator{
+			GetBalanceFunc: func(symbol string) ([]byte, error) {
+				return []byte(`testScript`), nil
+			},
+		}
+		invoker := &mocks.Invoker{
+			ScriptFunc: func(height uint64, script []byte, parameters []cadence.Value) (cadence.Value, error) {
+				return nil, mocks.DummyError
+			},
+		}
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			invoke:    invoker,
+		}
+
+		_, _, err := r.Balances(testBlockID, testAccount, testCurrencies)
+		assert.Error(t, err)
+	})
+}
+
+func TestRetriever_Block(t *testing.T) {
```

Comment:
I guess what we can do is to create separate functions to convert Flow types to Rosetta types, and remove all of that logic from the retriever.